### PR TITLE
Form failure status

### LINKED ISSUE #46: HTTP 400 is returned when it is the incorrect response code.
If deserializing the response data into a struct using `FromForm` fails, Rocket by default will return `400 BAD REQUEST`, regardless of the reason that deserialization failed. This is the incorrect response code, unless deserialization failed due to malformed JSON. It is clearly defined in [RFC 2616](https://tools.ietf.org/html/rfc2616#page-65) as "The request could not be understood by the server due to malformed syntax. The client SHOULD NOT repeat the request without modifications.". For failures due to missing fields, unrecognized fields, fields of the wrong type, or pretty much any failure that isn't the result of the body being syntactically incorrect for the declared content-type, [`422 UNPROCESSABLE ENTITY`](https://tools.ietf.org/html/rfc4918#page-78) should be used instead.

### PR DESCRIPTION
**Pull request by gsquire**
Fixes #46 

I may have misunderstood the scope of the issue but I am happy to update it if needed.

### CODE REVIEW

**Comment by SergioBenitez** [comment]
Just want to chime in and say I've not forgotten about you! Will have some time mid-next week to consider this thoroughly, and I'll get back to you then.

**Comment by gsquire** [comment]
It's no problem! I know this was slated for 0.2.

**Comment by SergioBenitez** [comment]
Alright! This is one of two remaining issues for 0.2!

So, looking at the [RFC, section 8.2.1](https://tools.ietf.org/html/rfc1866), we should return a `400` when any of the following occurs:

  1. A form string has no `=` at all or has no `=` after an `&`.
  2. The form string contains spaces.
  3. The form string contains unencoded URL characters.

The first one is pretty easy to detect but requires a change to the `FormItems` iterator. The second one means we have to check every string for spaces, which we can do efficiently if we change the `FormItems` decorator again. The last one...seems expensive, and I don't really know what the point of it is besides being able to send it in the URL.

Would be good to know what other frameworks do here. My intuition is that we should just check for 1 and ignore 2 and 3. This can be done by simply adding a method to the `FormItems` structure to see if there is any remaining, unparsed text. But even then: should Rocket just let this fly?

It's not clear to me what the correct approach is here.

**Comment by gsquire** [comment]
Thanks for doing the research. I don't know if the original issue author would want to weigh in at all, but I can adjust this PR when you settle on the approach.

File: lib/src/catcher.rs [hunk 1]
```diff
@@ -219,6 +219,8 @@ pub mod defaults {
                 teapot.", handle_418,
             421, "Misdirected Request", "The server cannot produce a response for this
                 request.", handle_421,
+            422, "Unprocessable Entity", "The server cannot process this request body.",
```
Comment (SergioBenitez) [inline lib/src/catcher.rs]: Change this to: "The request was well-formed but was unable to be followed due to semantic errors."

File: examples/forms/src/tests.rs [hunk 2]
```diff
@@ -40,6 +40,6 @@ fn test_bad_login() {
 #[test]
 fn test_bad_form() {
     // Mess with the form formatting.
-    test_login("Sergio&other=blah&", "password", 0, Status::BadRequest, None);
-    test_login("&&&===&", "password", 0, Status::BadRequest, None);
+    test_login("Sergio&other=blah&", "password", 0, Status::UnprocessableEntity, None);
```
Comment (SergioBenitez) [inline examples/forms/src/tests.rs]: This should be a `BadRequest`.

File: examples/forms/src/tests.rs [hunk 3]
```diff
@@ -40,6 +40,6 @@ fn test_bad_login() {
...
+    test_login("&&&===&", "password", 0, Status::UnprocessableEntity, None);
```
Comment (SergioBenitez) [inline examples/forms/src/tests.rs]: This should also be a `BadRequest`.

File: examples/forms/src/tests.rs [hunk 4]
```diff
@@ -40,6 +40,6 @@ fn test_bad_login() {
 #[test]
```
Comment (SergioBenitez) [inline examples/forms/src/tests.rs]: We should add some tests that should be `UnprocessableEntity`s.

**Comment by SergioBenitez** [comment]
Let's only worry about 1.

How do we go about implementing this? I see two possibilities:

  1. Change the `FromData` implementation for `Form` so that it validates the form string before passing it to the `FromForm` implementation and returns `BadRequest` on failed validation. This means that we iterate through the string twice when parsing forms.
  2. Change the `FromForm` trait so that it can inform the caller of a bad form string value somehow. Change that `FormItems` iterator so that it exposes a method to check if there is unparsed (i.e, invalid) input remaining. Change the `FromForm` derivation in codegen so that it calls this new method in `FormItems` and does the appropriate thing to inform the caller of a bad form string value. Finally, change the `Form` and `FromData` impl for `Form` to take this new value into account, returning the appropriate error.

I think **2** is the more "correct" approach, and it will be faster since we only go through the string once. But it's also a lot more complex. Maybe there's way to modify `FromForm` so that the changes are mostly backwards compatible? For instance, returning an `Option<Result<Self, Self::Error>>`, where `None` means that the form string was bad. Or something else, like changing the input type to be something mutable and inspecting it afterwards.

**Comment by SergioBenitez** [comment]
I went with a form of 2. I'm going to take your changes and merge them in with mine. :)

File: examples/forms/src/tests.rs (hunk 2 above)
Comment (SergioBenitez) [inline examples/forms/src/tests.rs]: This should be a BadRequest.

**Comment by gsquire** [comment]
Ok, thanks for the help. I will fix this up tonight when I get home.

**Comment by gsquire** [comment]
Are you going to add some tests for handling the 422 case? I guess I don't know how to do this without your changes :)

**Comment by SergioBenitez** [comment]
I will indeed. In the process of handling this now. Thanks so much for pushing these so quickly!

**Comment by SergioBenitez** [comment]
Merged in https://github.com/SergioBenitez/Rocket/commit/11e90f525f51c45eefb2acaef77414cfc24210b6. Thanks so much, again! :)

**Pull request closed without merging**